tests/test_references.py

```diff
@@ -185,6 +245,303 @@ async def test_deploy_end_to_end(
+    run_snippet("refs::plugins::takes_no_refs_dataclass_or_ref(refs::dc::create_no_refs_dataclass_reference())")
+    run_snippet("refs::plugins::takes_no_refs_dataclass_or_ref(refs::dc::NoRefsDataclass())")
+    run_snippet("refs::plugins::takes_mixed_refs_dataclass_or_ref(refs::dc::create_mixed_refs_dataclass_reference('hello'))")
+    run_snippet("refs::plugins::takes_mixed_refs_dataclass_or_ref(refs::dc::MixedRefsDataclass(maybe_ref_value='hello'))")
+    ## rejects other dataclasses / references to other dataclasses, except for inheritance
+    run_snippet("refs::plugins::takes_no_refs_dataclass_or_ref(refs::dc::MixedRefsDataclass(maybe_ref_value='hello'))")
+    run_snippet("refs::plugins::takes_no_refs_dataclass_or_ref(refs::dc::create_mixed_refs_dataclass_reference('hello'))")
+    with pytest.raises(
+        PluginTypeException, match=re.escape("Expected type: Reference[refs::dc::MixedRefsDataclass] | refs::dc::MixedRefsDataclass")
+    ):
+        run_snippet("refs::plugins::takes_mixed_refs_dataclass_or_ref(refs::dc::AllRefsDataclass(maybe_ref_value='hello'))")
+    with pytest.raises(
+        PluginTypeException, match=re.escape("Expected type: Reference[refs::dc::MixedRefsDataclass] | refs::dc::MixedRefsDataclass")
+    ):
+        run_snippet("refs::plugins::takes_mixed_refs_dataclass_or_ref(refs::dc::create_all_refs_dataclass_reference('hello'))")
+
+    # Scenario: inheritance on return type
+    ## declare generic, return specific
+    ### takes_no_refs_dataclass only accepts specific no_refs_dataclass.
+    ### We use it as an assertion that the other plugin returns the specific type in the DSL.
+    run_snippet("refs::plugins::takes_no_refs_dataclass(refs::plugins::inheritance_return_specific())")
+    ## declare generic reference, return specific reference
+    run_snippet("refs::plugins::takes_no_refs_dataclass_ref(refs::plugins::inheritance_return_specific_ref())")
+
+    # Scenario: plugin returns list attribute without reading elements, declares list[str] return
+    ## allowed if attribute has no references
+    run_snippet("refs::plugins::returns_entity_list(refs::ListContainer(value=['Hello', 'World!']))")
+    ## error on return validation if attribute has references
+    with pytest.raises(PluginTypeException, match="Return value .* has incompatible type\..*Expected type: string\[\]"):
+        run_snippet("refs::plugins::returns_entity_list(refs::ListContainer(value=['Hello', refs::create_string_reference('hello')]))")
+    ## allowed if plugin annotates reference in return type
+    run_snippet(
+        "refs::plugins::returns_entity_ref_list(refs::ListContainer(value=['Hello', refs::create_string_reference('hello')]))"
+    )
+
+    # Scenario: allow_reference_values() called on non-proxy list (e.g. list inside dataclass)
+    with pytest.raises(
+        ExternalException,
+        match="should only be called on inmanta instances, lists or dicts. Python lists and dicts do not need this wrapper",
```

Comment:
Why doesn't it just work on `any`? i.e. what does the increased restriction on the input gain us?